### PR TITLE
Adds the ability for you to fetch all children of an activity

### PR DESCRIPTION
**Pull request by m4rkmckenna**
Example 
GET http://localhost:8081/v1/activities/abc123/children/recurse

```
{
    "sCxsLgvm": {
        "id": "sCxsLgvm",
        "displayName": "invoking start[locations] on 1 node",
        "description": "",
        "entityId": "z81e0jl9j4",
        "entityDisplayName": "simple-app",
        "tags": [
            "SUB-TASK",
            {
                "type": "org.apache.brooklyn.core.test.entity.LocalManagementContextForTests"
            },
            {
                "wrappingType": "contextEntity",
                "entity": {
                    "type": "org.apache.brooklyn.api.entity.Entity",
                    "id": "z81e0jl9j4"
                }
            }
        ],
        "submitTimeUtc": 1466502440778,
        "startTimeUtc": 1466502440778,
        "endTimeUtc": 1466502441114,
        "currentStatus": "Completed",
        "result": [
            null
        ],
        "isError": false,
        "isCancelled": false,
        "children": [
            {
                "link": "/activities/WfYmdAuS",
                "metadata": {
                    "id": "WfYmdAuS",
                    "taskName": "start",
                    "entityId": "cmfr998e06",
                    "entityDisplayName": "simple-ent"
                }
            }
        ],
        "submittedByTask": {
            "link": "/activities/umIaqimp",
            "metadata": {
                "id": "umIaqimp",
                "taskName": "start",
                "entityId": "z81e0jl9j4",
                "entityDisplayName": "simple-app"
            }
        },
        "detailedStatus": "Completed after 336ms\n\nResult: [null]",
        "streams": {},
        "links": {
            "self": "/activities/sCxsLgvm",
            "children": "/activities/sCxsLgvm/children",
            "entity": "/applications/z81e0jl9j4/entities/z81e0jl9j4"
        }
    },
    "WfYmdAuS": {
        "id": "WfYmdAuS",
        "displayName": "start",
        "description": "Invoking effector start on simple-ent with parameters {locations=[LocalhostMachineProvisioningLocation{id=g1zmb43lie, name=localhost}]}",
        "entityId": "cmfr998e06",
        "entityDisplayName": "simple-ent",
        "tags": [
            "EFFECTOR",
            "SUB-TASK",
            {
                "wrappingType": "contextEntity",
                "entity": {
                    "type": "org.apache.brooklyn.api.entity.Entity",
                    "id": "cmfr998e06"
                }
            },
            {
                "type": "org.apache.brooklyn.core.test.entity.LocalManagementContextForTests"
            },
            {
                "wrappingType": "targetEntity",
                "entity": {
                    "type": "org.apache.brooklyn.api.entity.Entity",
                    "id": "cmfr998e06"
                }
            },
            {
                "entityId": "cmfr998e06",
                "effectorName": "start"
            }
        ],
        "submitTimeUtc": 1466502440778,
        "startTimeUtc": 1466502440778,
        "endTimeUtc": 1466502441114,
        "currentStatus": "Completed",
        "result": null,
        "isError": false,
        "isCancelled": false,
        "children": [
            {
                "link": "/activities/X9MqBujF",
                "metadata": {
                    "id": "X9MqBujF",
                    "taskName": "provisioning (localhost)",
                    "entityId": "cmfr998e06",
                    "entityDisplayName": "simple-ent"
                }
            },
            {
                "link": "/activities/ErGE1oPr",
                "metadata": {
                    "id": "ErGE1oPr",
                    "taskName": "pre-start",
                    "entityId": "cmfr998e06",
                    "entityDisplayName": "simple-ent"
                }
            },
            {
                "link": "/activities/ud8TXRaU",
                "metadata": {
                    "id": "ud8TXRaU",
                    "taskName": "start (processes)",
                    "entityId": "cmfr998e06",
                    "entityDisplayName": "simple-ent"
                }
            },
            {
                "link": "/activities/J1Kkj2xF",
                "metadata": {
                    "id": "J1Kkj2xF",
                    "taskName": "post-start",
                    "entityId": "cmfr998e06",
                    "entityDisplayName": "simple-ent"
                }
            }
        ],
        "submittedByTask": {
            "link": "/activities/sCxsLgvm",
            "metadata": {
                "id": "sCxsLgvm",
                "taskName": "invoking start[locations] on 1 node",
                "entityId": "z81e0jl9j4",
                "entityDisplayName": "simple-app"
            }
        },
        "detailedStatus": "Completed after 336ms\n\nNo return value (null)",
        "streams": {},
        "links": {
            "self": "/activities/WfYmdAuS",
            "children": "/activities/WfYmdAuS/children",
            "entity": "/applications/z81e0jl9j4/entities/cmfr998e06"
        }
    },
    "X9MqBujF": {
        "id": "X9MqBujF",
        "displayName": "provisioning (localhost)",
        "description": "",
        "entityId": "cmfr998e06",
        "entityDisplayName": "simple-ent",
        "tags": [
            "SUB-TASK",
            {
                "wrappingType": "contextEntity",
                "entity": {
                    "type": "org.apache.brooklyn.api.entity.Entity",
                    "id": "cmfr998e06"
                }
            },
            {
                "type": "org.apache.brooklyn.core.test.entity.LocalManagementContextForTests"
            }
        ],
        "submitTimeUtc": 1466502440782,
        "startTimeUtc": 1466502440782,
        "endTimeUtc": 1466502440799,
        "currentStatus": "Completed",
        "result": {
            "type": "org.apache.brooklyn.api.location.Location",
            "id": "fe4jz35nvm"
        },
        "isError": false,
        "isCancelled": false,
        "children": [],
        "submittedByTask": {
            "link": "/activities/WfYmdAuS",
            "metadata": {
                "id": "WfYmdAuS",
                "taskName": "start",
                "entityId": "cmfr998e06",
                "entityDisplayName": "simple-ent"
            }
        },
        "detailedStatus": "Completed after 17ms\n\nResult: SshMachineLocation[localhost:null@chimera.local/192.168.4.161:22(id=fe4jz35nvm)]",
        "streams": {},
        "links": {
            "self": "/activities/X9MqBujF",
            "children": "/activities/X9MqBujF/children",
            "entity": "/applications/z81e0jl9j4/entities/cmfr998e06"
        }
    },
########################
# RESPONSE TRUNCATED #
########################
}
```

### CODE REVIEW

File: rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java [hunk 1]
```diff
@@ -57,6 +57,18 @@ public TaskSummary get(
...
     @GET
+    @Path("/{task}/children/recurse")
+    @ApiOperation(
+            value = "Fetch all child tasks details",
+            response = org.apache.brooklyn.rest.domain.TaskSummary.class,
+            responseContainer = "map")
+    @ApiResponses(value = {
+            @ApiResponse(code = 404, message = "Could not find task")
+    })
+    public Map<String,TaskSummary> getAllChildrenAsMap(
```
Comment (geomacy) [inline rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java]: Or 'getDescendents' to clarify that it's all children and their children.

Comment (m4rkmckenna) [inline rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java]: I like get descendants ... I'll change shortly

**Comment by geomacy** [comment]
Looks good to me. (Musing: At first sight it looked odd to return a list of children for `/children` and a map for `/children/recurse`, but on further thought the map _is_ appropriate for the latter case, and the list for the former, so fair enough.)

File: rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java [hunk 2]
```diff
@@ -57,6 +57,18 @@ public TaskSummary get(
             @ApiParam(value = "Task ID", required = true) @PathParam("task") String taskId);
 
     @GET
+    @Path("/{task}/children/recurse")
+    @ApiOperation(
+            value = "Fetch all child tasks details",
+            response = org.apache.brooklyn.rest.domain.TaskSummary.class,
```
Comment (neykov) [inline rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java]: The response is a `Map`, not an instance of the class.

Comment (neykov) [inline rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java]: Oh there's a `responseContainer`, never mind.

File: rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java [hunk 3]
```diff
@@ -57,6 +57,18 @@ public TaskSummary get(
...
+            responseContainer = "map")
```
Comment (neykov) [inline rest/rest-api/src/main/java/org/apache/brooklyn/rest/api/ActivityApi.java]: Looks like that's not a valid value - http://docs.swagger.io/swagger-core/apidocs/com/wordnik/swagger/annotations/ApiOperation.html#responseContainer()

**Comment by neykov** [comment]
Looks good, apart from the documentation quirks (not sure how it should look like though).

What's the use case for this API? Shouldn't we include the top-level task in the response as well? I expected that before reading the code.
What about returning all the tasks of an entity instead?

**Comment by m4rkmckenna** [comment]
@neykov 
> What's the use case for this API? 

When debugging you are usually only interested in one task being able to fetch the full tree makes debugging much easier as you do not have walk the full tree of data fetching children at each level 

> What about returning all the tasks of an entity instead?

Returning the full task tree of an entity could be a massive amount of data.

Re Documentation quirks ... My bad i thought swagger supported responseContainer Map since 2.0. I will amend shortly

**Comment by neykov** [comment]
Thanks for the update, merging.

**Pull request merged**